### ISSUE TITLE
mingw: multiple definition of `_time64'

### ISSUE THREAD

**Issue by kroeckx** [openssl/openssl#3280]
We're getting this on x86_64-w64-mingw32-gcc:

```
/usr/lib/gcc/x86_64-w64-mingw32/4.8/../../../../x86_64-w64-mingw32/lib/libmsvcrt.a(dlptbs00688.o):(.text+0x0): multiple definition of `_time64'
fuzz/client.o:client.c:(.text+0x20): first defined here
```

I'm confused about what that means, is this on win64, but creating a 32 bit binary?

**Comment by mattcaswell** [comment]
> I'm confused about what that means, is this on win64, but creating a 32 bit binary?

This is cross compiling a Win64 target on a Linux host using mingw32 (its called that whether you are using it for 32-bit or 64-bit - no idea why).

I can't replicate this - either in native Win64 with mingw32, or cross compiling from Linux.

**Comment by levitte** [comment]
Hmmm, I think I did stumble on this not too long ago.  Will re-check

**Comment by dot-asm** [comment]
> (its called that whether you are using it for 32-bit or 64-bit - no idea why)

Because there is _time64 even on 32-bit Windows. To recap, idea behind this trick is to override library time function. Since we don't know for sure which one is actually called on Windows elsewhere, both are implemented. One should also keep in mind (and this was also discussed) that in shared (Windows) case it wouldn't do the trick anyway. To summarize, the is a lot of ifs and buts and it's quite opportunistic approach. But it was conscious choice with rationale "if doesn't work out, it's not end of the world". It ought to be linker-specific thing, and if one fails, one can as well just omit _time64 implementation, It's not bigger loss since fuzzer tests would be performed, and wouldn't be meaningless even if not performed with fixed time...

**Comment by sav-ix** [comment]
Hello,

Got this error while testing (recently introduced) static-only builds using mingw-w64:
```
sh-4.4$ ./Configure --prefix="$PWD/../GCC64RT" threads no-deprecated no-shared mingw64 && make
[snip]
make[2]: Entering directory '/c/libOPENSSL-1.1.1-dev/build'
( :; LIBDEPS="${LIBDEPS:--L. -lssl -L. -lcrypto -lws2_32 -lgdi32 -lcrypt32 }"; LDCMD="${LDCMD:-gcc}"; LDFLAGS="${LDFLAGS:--DDSO_WIN32 -DNDEBUG -DOPENSSL_THREADS -DOPENSSL_NO_DYNAMIC_ENGINE -DOPENSSL_PIC -DOPENSSL_IA32_SSE2 -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_MONT5 -DOPENSSL_BN_ASM_GF2m -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DRC4_ASM -DMD5_ASM -DAES_ASM -DVPAES_ASM -DBSAES_ASM -DGHASH_ASM -DECP_NISTZ256_ASM -DPADLOCK_ASM -DPOLY1305_ASM -DOPENSSL_API_COMPAT=0x10100000L -DOPENSSLDIR="\"/c/libOPENSSL-1.1.1-dev/build/../GCC64RT/ssl\"" -DENGINESDIR="\"/c/libOPENSSL-1.1.1-dev/build/../GCC64RT/lib/engines-1_1\"" -DL_ENDIAN -DWIN32_LEAN_AND_MEAN -DUNICODE -D_UNICODE -m64 -Wall -O3 -D_MT   }"; LIBPATH=`for x in $LIBDEPS; do echo $x; done | sed -e 's/^ *-L//;t' -e d | uniq`; LIBPATH=`echo $LIBPATH | sed -e 's/ /:/g'`; echo LD_LIBRARY_PATH=$LIBPATH:$LD_LIBRARY_PATH ${LDCMD} ${LDFLAGS} -o ${APPNAME:=fuzz/client-test.exe} fuzz/client.o fuzz/test-corpus.o ${LIBDEPS}; LD_LIBRARY_PATH=$LIBPATH:$LD_LIBRARY_PATH ${LDCMD} ${LDFLAGS} -o ${APPNAME:=fuzz/client-test.exe} fuzz/client.o fuzz/test-corpus.o ${LIBDEPS} )
LD_LIBRARY_PATH=.: gcc -DDSO_WIN32 -DNDEBUG -DOPENSSL_THREADS -DOPENSSL_NO_DYNAMIC_ENGINE -DOPENSSL_PIC -DOPENSSL_IA32_SSE2 -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_MONT5 -DOPENSSL_BN_ASM_GF2m -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DRC4_ASM -DMD5_ASM -DAES_ASM -DVPAES_ASM -DBSAES_ASM -DGHASH_ASM -DECP_NISTZ256_ASM -DPADLOCK_ASM -DPOLY1305_ASM -DOPENSSL_API_COMPAT=0x10100000L -DOPENSSLDIR="/c/libOPENSSL-1.1.1-dev/build/../GCC64RT/ssl" -DENGINESDIR="/c/libOPENSSL-1.1.1-dev/build/../GCC64RT/lib/engines-1_1" -DL_ENDIAN -DWIN32_LEAN_AND_MEAN -DUNICODE -D_UNICODE -m64 -Wall -O3 -D_MT -o fuzz/client-test.exe fuzz/client.o fuzz/test-corpus.o -L. -lssl -L. -lcrypto -lws2_32 -lgdi32 -lcrypt32
C:/ProgramData/Mingw64/bin/../lib/gcc/x86_64-w64-mingw32/6.3.0/../../../../x86_64-w64-mingw32/lib/../lib/libmsvcrt.a(diafs00791.o):(.text+0x0): multiple definition of `_time64'
fuzz/client.o:client.c:(.text+0x20): first defined here
collect2.exe: error: ld returned 1 exit status
make[2]: *** [Makefile.shared:187: link_app.] Error 1
make[2]: Leaving directory '/c/libOPENSSL-1.1.1-dev/build'
make[1]: *** [Makefile:6629: fuzz/client-test.exe] Error 2
make[1]: Leaving directory '/c/libOPENSSL-1.1.1-dev/build'
make: *** [Makefile:133: all] Error 2
```
Reproduced for `Release+Static` builds only.

Environment:
- Windows 10 x64,
- mingw-w64 x86_64 6.3.0,
- MSYS2 20161025,
- OpenSSL-1.1.1-dev (19044d3c22ffcb711e62db11634476917a0a58d9).

Alexander

**Comment by levitte** [comment]
So here's what I find out about `_time64`:

```
$ nm -Pg /usr/x86_64-w64-mingw32/lib/libmsvcrt.a | grep -A5 libmsvcrts00692.o
/usr/x86_64-w64-mingw32/lib/libmsvcrt.a[libmsvcrts00692.o]:
_head_lib64_libmsvcrt_a U
__imp__time64 I 0000000000000000 
_time64 T 0000000000000000 
/usr/x86_64-w64-mingw32/lib/libmsvcrt.a[libmsvcrts00691.o]:
_head_lib64_libmsvcrt_a U
```

I was thinking that `__imp__time64` might be a reason that this particular object file (libmsvcrts00692.o) got sucked in, so tried reimplementing that symbols as well in `fuzz/client.c`...  unfortunately, that made no difference...

**Comment by mattcaswell** [comment]
Can we just detect that we are running in mingw and not define the symbol in that case?

**Comment by levitte** [comment]
Wasn't the point of this time exercise to "override" the standard time function to be able to get a constant time value?  If we avoid it entirely, doesn't that mean the fuzzing that involves time will risk failure on mingw, without actual cause other than time being somewhat unpredictable?

**Comment by dot-asm** [comment]
As already said multiple times. It's opportunistic approach (has to be) and if it doesn't work some where, it's not the end of the world. It also makes lesser sense to try to tackle every possible situation with #if-spaghetti, it's not worth it. This is because there are cases (once again, shared Windows build) where *nothing* [in source code alone] would do the trick. For this reason, since it's now shown to be not working universally, it should be reduced to
```
#if !defined(_WIN32)
time_t time(time_t *t) TIME_IMPL(t)
#endif
```
That's it. And with time we should introduce OPENSSL_time or something that would return 64-bit time, preferably in alternative units (cross-ref to #3105). At that point it would be possible to solve this not-really-a-problem for all supported platforms. But meanwhile, just omit...

**Comment by dot-asm** [comment]
Addressed in #3320.
